File: 03_Backend/app/models/qr_code.py
"""QrCode — QR code entity (``qr_code``).

Business purpose
----------------
QR codes are the primary identification mechanism for interns and visitors.
Each QR code has a status lifecycle: ``ACTIF`` → ``EXPIRE`` / ``REVOQUE``.

* Interns get one active QR code that expires on their internship end date.
* Visitors get one QR code that expires at 23:59 on their visit date.
* Employees never receive QR codes — they use Face Recognition.

Every QR code stores a **SHA-256 hash** of the raw token, so the database
never holds the plaintext token.  The raw token is returned only at
generation time and is embedded in the PNG image.
"""

from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import BaseModel


class QrCode(BaseModel):
    __tablename__ = "qr_code"

    qr_hash: Mapped[str] = mapped_column(
        String(255),
        unique=True,
        index=True,
    )

    proprietaire_uuid: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("utilisateur.uuid"),
        index=True,
    )

    type_proprietaire: Mapped[str] = mapped_column(String(20))

    statut: Mapped[str] = mapped_column(
        String(20),
        default="ACTIF",
        index=True,
    )

    date_expiration: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
    )

    cree_par_uuid: Mapped[str | None] = mapped_column(
        String(36),
        ForeignKey("utilisateur.uuid"),
        default=None,
    )

    date_revocation: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        default=None,
    )

    revoque_par_uuid: Mapped[str | None] = mapped_column(
        String(36),
        ForeignKey("utilisateur.uuid"),
        default=None,
    )

    motif_revocation: Mapped[str | None] = mapped_column(
        String(100),
        default=None,
    )

    derniere_validation: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        default=None,
    )

    nombre_validations: Mapped[int] = mapped_column(
        default=0,
    )

    metadata_json: Mapped[str | None] = mapped_column(
        Text,
        default=None,
    )

    def __repr__(self) -> str:
        return (
            f"<QrCode id={self.id} statut={self.statut} "
            f"proprietaire={self.proprietaire_uuid} type={self.type_proprietaire}>"
        )

File: mobile_app/lib/features/admin/qr/domain/entities/qr_code.dart
class QrCode {
  final int id;
  final String uuid;
  final String qrHash;
  final String proprietaireUuid;
  final String typeProprietaire;
  final String statut;
  final DateTime dateExpiration;
  final String? creeParUuid;
  final DateTime? dateRevocation;
  final String? revoqueParUuid;
  final String? motifRevocation;
  final DateTime? derniereValidation;
  final int nombreValidations;
  final String? metadataJson;
  final String? qrBase64;
  final String? proprietaireNom;
  final String? proprietairePrenom;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const QrCode({
    required this.id,
    required this.uuid,
    required this.qrHash,
    required this.proprietaireUuid,
    required this.typeProprietaire,
    required this.statut,
    required this.dateExpiration,
    this.creeParUuid,
    this.dateRevocation,
    this.revoqueParUuid,
    this.motifRevocation,
    this.derniereValidation,
    this.nombreValidations = 0,
    this.metadataJson,
    this.qrBase64,
    this.proprietaireNom,
    this.proprietairePrenom,
    this.createdAt,
    this.updatedAt,
  });

  String get proprietaireFullName {
    if (proprietaireNom != null && proprietairePrenom != null) {
      return '${proprietairePrenom} ${proprietaireNom}';
    }
    return proprietaireUuid;
  }

  bool get isActive => statut == 'ACTIF';

  bool get isExpired => statut == 'EXPIRE';

  bool get isRevoked => statut == 'REVOQUE';

  String get typeLabel {
    switch (typeProprietaire) {
      case 'STAGIAIRE':
        return 'Stagiaire';
      case 'VISITEUR':
        return 'Visiteur';
      default:
        return typeProprietaire;
    }
  }
}

class PaginatedQrCodes {
  final List<QrCode> items;
  final int total;
  final int page;
  final int pageSize;
  final int totalPages;

  const PaginatedQrCodes({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
    required this.totalPages,
  });

  bool get hasMore => page < totalPages;
}
